### PR TITLE
fix(cra-metrics): Count all users in metrics alerts [INGEST-1330 INGEST-1371]

### PR DESCRIPTION
**Pull request by jjbayer**
As in https://github.com/getsentry/sentry/pull/34858, remove the `session.status=init` filter to get a more accurate total number of users.

Use conditional aggregates in order to get both the total user count _and_ the number of crashed users in the same snuba query.

To maintain compatibility until existing subscriptions have been migrated, make the subscription processor able to handle both the old and the new format. The actual migration of existing subscriptions will be in a separate PR.

### CODE REVIEW

File: src/sentry/snuba/entity_subscription.py [hunk 1]
```diff
@@ -375,13 +357,50 @@ def aggregate_query_results(
...
+                f"sumIf(value, equals({self.session_status}, {session_status_crashed}))",
+                None,
+                "crashed",
+            ],
+        ]
+
+    def get_extra_conditions(self) -> List[List[Any]]:
+        crashed = resolve(self.org_id, "crashed")
+        init = resolve(self.org_id, "init")
+        return [[self.session_status, "IN", [crashed, init]]]
```
Comment (jjbayer) [inline src/sentry/snuba/entity_subscription.py:380]: For the *session* count, we still use `status=init` as the total count, so it makes sense to keep the session status filter.

File: src/sentry/snuba/entity_subscription.py [hunk 2]
```diff
@@ -375,13 +357,50 @@ def aggregate_query_results(
...
+            ],
+            [
+                f"uniqIf(value, equals({self.session_status}, {session_status_crashed}))",
+                None,
+                "crashed",
+            ],
+        ]
+
+    def get_extra_conditions(self) -> List[List[Any]]:
+        return []
```
Comment (jjbayer) [inline src/sentry/snuba/entity_subscription.py:403]: For the *user* count, we now query the metric over all statuses, so we cannot filter down to status.

File: tests/sentry/incidents/test_subscription_processor.py [hunk 3]
```diff
@@ -2121,37 +2114,6 @@ def test_multiple_threshold_resolve_is_reset_when_count_is_lower_than_min_thresh
         self.assert_trigger_exists_with_status(incident, trigger, TriggerStatus.ACTIVE)
         self.assert_action_handler_called_with_actions(incident, [])
 
-    def test_ensure_case_when_no_metrics_index_not_found_is_handled_gracefully(self):
```
Comment (jjbayer) [inline tests/sentry/incidents/test_subscription_processor.py:2124]: The total `count` and `crashed` count now have explicit aliases, and we no longer group by `session.status`, so there will not be any unexpected values for that tag in the query result.

File: src/sentry/snuba/entity_subscription.py [hunk 4]
```diff
@@ -375,13 +357,50 @@ def aggregate_query_results(
...
-    aggregation_func: str = "sum"
+
+    def get_aggregations(self) -> List[List[Optional[str]]]:
+        session_status_crashed = resolve(self.org_id, "crashed")
+        session_status_init = resolve(self.org_id, "init")
+        return [
+            [
+                f"sumIf(value, equals({self.session_status}, {session_status_init}))",
+                None,
+                "count",
```
Comment (ahmedetefy) [inline src/sentry/snuba/entity_subscription.py:368]: Really nit picking here so feel free to ignore
Why is this called `count` and not `total` ?
Found it to be a bit confusing

**Review by ahmedetefy — APPROVED** [review]
This lgtm! 
I would wait to merge this until maybe @wedamija takes a look at this as this changes the subscription queries for crash rate alerts over metrics, and workflow owns that now

File: src/sentry/snuba/entity_subscription.py (hunk 4 above)
Comment (jjbayer) [reply src/sentry/snuba/entity_subscription.py:368]: I agree that `total` would be a better name. I noticed we overwrite the alias with `"count"` here: https://github.com/getsentry/sentry/blob/905088db962c02b2e4f38dde3724850048b25734/src/sentry/incidents/logic.py#L377
so I thought we might as well use it as the original alias.

**Review by wedamija — CHANGES_REQUESTED** [review]
Do we need to recreate existing subscriptions here? It looks like we're changing how we create the aggregations for these subscriptions. So this means that any current subscriptions will still have the old format.

If that's the case, then we'll need some kind of migrations procedure, and will probably need to leave the old code place while we transition.

**Comment by jjbayer** [comment]
> Do we need to recreate existing subscriptions here? It looks like we're changing how we create the aggregations for these subscriptions. So this means that any current subscriptions will still have the old format.
> 
> If that's the case, then we'll need some kind of migrations procedure, and will probably need to leave the old code place while we transition.

@wedamija I see, so if we want to update existing subscriptions we basically need to call `update_alert_rule` on all of them? Has a migration like this been done before?

If I understand correctly, we need the subscriptions processor to handle both formats, regardless of whether we update old subscriptions or not, because there will always be a period of time where both formats exist.

**Comment by wedamija** [comment]
> > Do we need to recreate existing subscriptions here? It looks like we're changing how we create the aggregations for these subscriptions. So this means that any current subscriptions will still have the old format.
> > If that's the case, then we'll need some kind of migrations procedure, and will probably need to leave the old code place while we transition.
> 
> @wedamija I see, so if we want to update existing subscriptions we basically need to call `update_alert_rule` on all of them? Has a migration like this been done before?

Yep, there are scripts that do this fairly easily. I would recommend doing this in a migration so that it goes to all customers. Here's an example: https://github.com/getsentry/getsentry/blob/89ff1453be755ddef31f2b99de09bd03badeb25e/getsentry/migrations/0141_migrate_sessions_subs_to_metrics.py#L116-L115

So you'd want to make a new migration in `sentry` and copy that logic. Then modify the queries to just select the specific rows you want to update.

> If I understand correctly, we need the subscriptions processor to handle both formats, regardless of whether we update old subscriptions or not, because there will always be a period of time where both formats exist.

Yeah, we will need to have a transition state that can handle both the new and old format. Then if we do a migration, once it has run we can remove the deprecated format handling code.

**Comment by jjbayer** [comment]
> Yeah, we will need to have a transition state that can handle both the new and old format. Then if we do a migration, once it has run we can remove the deprecated format handling code.

@wedamija thanks for the guidance! I've implemented the transition state on this PR, and will follow up with a separate PR for the migration.

**Review by wedamija — APPROVED** [review]

**Pull request merged**